nexus/db-queries/src/db/datastore/region.rs

```diff
@@ -118,44 +119,67 @@ impl DataStore {
-        // favoring datasets with the smallest existing (summed) region
-        // sizes. Basically, "pick the datasets with the smallest load first".
-        //
-        // Longer-term, we should consider:
-        // - Storage size + remaining free space
-        // - Sled placement of datasets
-        // - What sort of loads we'd like to create (even split across all disks
-        // may not be preferable, especially if maintenance is expected)
-
         let block_size =
             self.get_block_size_from_disk_source(opctx, &disk_source).await?;
         let (blocks_per_extent, extent_count) =
             Self::get_crucible_allocation(&block_size, size);
 
-        let dataset_and_regions: Vec<(Dataset, Region)> =
-            crate::db::queries::region_allocation::RegionAllocate::new(
-                volume_id,
-                block_size.into(),
-                blocks_per_extent,
-                extent_count,
-            )
-            .get_results_async(self.pool())
-            .await
-            .map_err(|e| crate::db::queries::region_allocation::from_pool(e))?;
-
-        Ok(dataset_and_regions)
+        // Right now the SQL queries to allocate regions will shuffle all the
+        // datasets and pick 3. However, if a zpool has multiple datasets, it
+        // may pick 2 datasets on the same zpool. We do not want this, because
+        // then you have 2 datasets on the same disk, and now you don't have
+        // redundancy. We can't deal with that in the SQL query, but we can
+        // *detect* it. So we detect it.
+        //
+        // When a query fails because it picked 2 zpools from the same dataset,
+        // but there were 3 valid zpools it could have picked from,
+        // can_be_retried(error) will return true. It is highly likely that if
+        // we simply retry the query with a different shuffle-ordering, we will
+        // get a valid result. However, we should no retry forever, in the event
+        // that something else is going on, so we limit it.
+        //
+        // In theory we could also fail on the criteria of picking datasets from
+        // the same sled, and retry until we get 3 different sleds.
+        //
+        // It does seem like a bit of a hack, doesn't it?
+
+        const RETRIES: usize = 10;
+        let mut times_tried = 0;
+
+        // If the strat specified a certain seed, we increment the seed with
+        // each retry, so that the retry-shuffles are still based on the input
+        // seed in some fashion.
+        let strats: Vec<RegionAllocationStrategy> = match allocation_strategy {
+            RegionAllocationStrategy::Random(Some(seed)) => (*seed
+                ..=*seed + RETRIES as u128)
+                .map(|seed| RegionAllocationStrategy::Random(Some(seed)))
+                .collect(),
+            strat => (0..=RETRIES).map(|_| strat.clone()).collect(),
+        };
+        loop {
```

Comment:
Chatted in a meet about this -- let's take out the retry loop here, and make this an explicit error. This condition should only happen "when we pick two datasets from the same zpool", which like, should not happen on any of the configurations we currently support. Throwing an error here is fine, but we shouldn't feel compelled to retry.

(Separately, if we follow-up by provisioning from distinct sleds, this issue will be moot anyway)

Reply:
retry removed!